docs: Better documentation of trusting SSL certs

bmingles opened the pull request:
DH-18191: Better documentation of trusting SSL certs

NodeJS doesn't use the OS cert store to determine which signing certificates it trusts. Instead, it has a static list of trusted certs included in the install. For certificates not in this list, there is a `NODE_EXTRA_CA_CERTS` environment variable that can be set to tell NodeJS about additional certs to trust. This PR updates the docs with instructions of how to set this.

> Note there is some indication that extensions may eventually allow using Chrome's `fetch`. If / when that happens, it might be possible to use that for Deephaven extension instead of our current grpc transport implementation that uses `node:http2` module. This could allow extensions to use the OS cert store instead of this env variable.
https://github.com/microsoft/vscode/issues/12588#issuecomment-2111861237

arman-ddl commented on README.md:
````diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+   Windows
+
+   ```sh
+   setx NODE_EXTRA_CA_CERTS=C:\Path\To\cert.pm
````
I'm pretty sure setx uses a space and not equal sign between variable name and value.

arman-ddl commented on README.md:
````diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+   
+   On Mac / Linux
+   ```sh
+   export NODE_EXTRA_CA_CERTS=/path/to/cert.pm
````
1. Would be nice to provide Unix instructions that the user only needs to run a single time like we do with Windows (`setx`).
2. Should .pm suffix be .pem?

bmingles replied:
Updated

bmingles commented on README.md:
````diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+   setx NODE_EXTRA_CA_CERTS=C:\Path\To\cert.pm
+   ```
+   > Note that paths in env variables should not be wrapped in quotes on Windows.
+1. Start VS Code in the environment where you set the `NODE_EXTRA_CA_CERTS` variable.
````
Unix can use either as well. Updated

bmingles commented on README.md:
````diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+   Windows
+
+   ```sh
+   setx NODE_EXTRA_CA_CERTS=C:\Path\To\cert.pm
````
yes

mofojed commented on README.md:
```diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+Deephaven servers using self-signed certificates or internal CA's will require configuring VS Code to trust the signing certificate.
+
+1. Save the signing certificate in PEM format somewhere on the machine running VS Code. Multiple certificates can be concatenated together in the same file if there are multiple certs that need to be configured.
+1. Set the `NODE_EXTRA_CA_CERTS` environment variable to the path of the signing certificate.
```
```suggestion
2. Set the `NODE_EXTRA_CA_CERTS` environment variable to the path of the signing certificate.
```

bmingles replied:
@mofojed Is this change necessary? Markdown rendererer "should" auto number these.

bmingles replied:
You can see the output in the branch:
https://github.com/deephaven/vscode-deephaven/tree/DH-18191-2_logs-and-docs?tab=readme-ov-file#ssl-certificates

mofojed replied:
Yes it does auto number them, was a nitpick.

mofojed commented on README.md:
````diff
@@ -61,6 +58,29 @@ Enterprise servers can be configured via the `"deephaven.enterpriseServers"` set
+   setx NODE_EXTRA_CA_CERTS C:\Path\To\cert.pem
+   ```
+   > Note that paths in env variables should not be wrapped in quotes on Windows.
+1. Start VS Code in a shell that has the `NODE_EXTRA_CA_CERTS` variable set.
````
```suggestion
3. Start VS Code in a shell that has the `NODE_EXTRA_CA_CERTS` variable set.
```

bmingles replied:
Same here, Markdown supports numbering as 1. 1. 1. ... and auto increments. Makes it easier to insert / re-order in the future without modifying every line.

mofojed reviewed (changes requested):
Add a little more context to the PR description.

In particular, I think it's important we put the ticket number for VS Code tracking this.

In their documentation: https://code.visualstudio.com/docs/setup/network#_legacy-proxy-server-support

> Extensions don't benefit yet from the same proxy support that VS Code supports. You can follow this issue's development in [GitHub](https://github.com/microsoft/vscode/issues/12588).

And the issue itself has some information on their efforts: https://github.com/microsoft/vscode/issues/12588

bmingles commented:
> Add a little more context to the PR description.
> 
> In particular, I think it's important we put the ticket number for VS Code tracking this.
> 
> In their documentation: https://code.visualstudio.com/docs/setup/network#_legacy-proxy-server-support
> 
> > Extensions don't benefit yet from the same proxy support that VS Code supports. You can follow this issue's development in [GitHub](https://github.com/microsoft/vscode/issues/12588).
> 
> And the issue itself has some information on their efforts: [microsoft/vscode#12588](https://github.com/microsoft/vscode/issues/12588)

@mofojed I'm not sure this is directly related to proxy support. It's a NodeJs nuance of trusting SSL certificates. Even if there is no proxy involved, a signing cert that is not included in the NodeJS install won't be trusted without this config.

mofojed commented:
@bmingles Okay I found one comment on that linked issue though that seems interesting: https://github.com/microsoft/vscode/issues/12588#issuecomment-2111861237

> Work is under way to make Chromium's implementation of fetch available to extensions. This issue is unlikely to be closed soon though, as many extensions are using Node's https module for which we have only partial support for proxies. Only proxies without auth and with Kerberos auth are supported at the moment.

Sounds like if they provided that `fetch`, we could just use that and be good

bmingles commented:
@mofojed I added that comment to the PR description. I also added a comment to that issue to see if anyone can provide a more direct link to Chrome `fetch` support. I'll update the PR description if / when I get a response.
